bpo-31945: Configurable blocksize in HTTP(S)Connection

nirs opened the pull request:
blocksize was hardcoded to 8192, preventing efficient upload when using
file-like body. Add blocksize argument to __init__, so users can
configure the blocksize to fit their needs.

I tested this uploading data from /dev/zero to a web server dropping the
received data, to test the overhead of the HTTPConnection.send() with a
file-like object.

Here is an example 10g upload with the default buffer size (8192):

$ time ~/src/cpython/release/python upload-httplib.py 10 https://localhost:8000/
Uploaded 10.00g in 17.53 seconds (584.00m/s)

real	0m17.574s
user	0m8.887s
sys	0m5.971s

Same with 512k blocksize:

$ time ~/src/cpython/release/python upload-httplib.py 10 https://localhost:8000/
Uploaded 10.00g in 6.60 seconds (1551.15m/s)

real	0m6.641s
user	0m3.426s
sys	0m2.162s

In real world usage the difference will be smaller, depending on the
local and remote storage and the network.

See https://github.com/nirs/http-bench for more info.

<!-- issue-number: bpo-31945 -->
https://bugs.python.org/issue31945
<!-- /issue-number -->

nirs committed:
bpo-31945: Configurable blocksize in HTTP(S)Connection

blocksize was hardcoded to 8192, preventing efficient upload when using
file-like body. Add blocksize argument to __init__, so users can
configure the blocksize to fit their needs.

I tested this uploading data from /dev/zero to a web server dropping the
received data, to test the overhead of the HTTPConnection.send() with a
file-like object.

Here is an example 10g upload with the default buffer size (8192):

$ time ~/src/cpython/release/python upload-httplib.py 10 https://localhost:8000/
Uploaded 10.00g in 17.53 seconds (584.00m/s)

real	0m17.574s
user	0m8.887s
sys	0m5.971s

Same with 512k blocksize:

$ time ~/src/cpython/release/python upload-httplib.py 10 https://localhost:8000/
Uploaded 10.00g in 6.60 seconds (1551.15m/s)

real	0m6.641s
user	0m3.426s
sys	0m2.162s

In real world usage the difference will be smaller, depending on the
local and remote storage and the network.

See https://github.com/nirs/http-bench for more info.

nirs committed:
Add news entry

nirs committed:
Update module documentation

vstinner commented on Doc/library/http.client.rst:
```diff
@@ -42,6 +43,8 @@ The module provides the following classes:
    (if it is not given, the global default timeout setting is used).
    The optional *source_address* parameter may be a tuple of a (host, port)
    to use as the source address the HTTP connection is made from.
+   The optional *blocksize* parameter sets the buffer size used to send a
```
Please add "in bytes" to size.

nirs replied:
Thanks for the reivew, will update in the next version.

I'll check how to add a test for this.

vstinner reviewed (commented):
Would it be possible to write a test for it, maybe using a mock socket?

vadmium commented on Lib/http/client.py:
```diff
@@ -825,9 +825,10 @@ def _get_content_length(body, method):
+                 source_address=None, blocksize=8192):
         self.timeout = timeout
         self.source_address = source_address
+        self.blocksize = blocksize
```
I suggest either name the attribute with an underscore (“_blocksize”) or document it as a public attribute.

nirs replied:
I'll document that this is a public attribute, since a user of the connection may like to modify it.

For example, user code may want to check if a connection supports changing the block size, and ``hasattr(conn, "blocksize")`` is nicer to use compared with checking ``__init__`` signature.

vadmium commented:
For testing, just check the parameter passed to the _read_ method, and use whatever socket or server the other tests use.

nirs committed:
Improve documentation

- Note that buffer size is "in bytes" (Victor)
- Document the new public "blocksize" attribute (Martin)

vadmium commented on Doc/library/http.client.rst:
```diff
@@ -395,6 +401,13 @@ also send your request step by step, by using the four functions below.
    called.
 
 
+.. attribute:: HTTPConnection.blocksize
```
This is in a list of “four functions” for low-level requests. Perhaps move it up into the main list, and adjust or remove “the following methods”.

nirs replied:
Ok, will move up before the low level method section.

nirs committed:
Fix new docs

- Move blocksize attribute documentaion before the low level methods
  section (Martin)

nirs committed:
Add test for new blocksize parameter

Test that the send loop uses the configured block size when sending a
readable object.

nirs committed:
Test block size both in send() and request() flows

We have two different code paths using blocksize - send() and
_read_readable(). Add a test for the second code path.

nirs committed:
Refine news entry

vstinner commented on Misc/NEWS.d/next/Library/2017-11-05-01-17-12.bpo-31945.TLPBtS.rst:
```diff
@@ -0,0 +1,2 @@
+Add Configurable blocksize to ``HTTPConnection`` and ``HTTPSConnection``
```
I propose ``Add configurable *blocksize* parameter to (...)``.

vstinner reviewed (approved):
LGTM. Thank you for the added unit tests.

You might add an entry to Doc/whatsnew/3.7.rst

nirs committed:
Refine rst formatting in news entry

nirs commented:
> You might add an entry to Doc/whatsnew/3.7.rst
@haypo What kind of info should be in whatsnew?

vstinner commented:
>  @haypo What kind of info should be in whatsnew?

Basically, copy/paste your NEWS entry in the "Improved Modules" section: it seems like you have to add a new "socket" sub-section.

nirs committed:
Add entry to Doc/whatsnew/3.7.rst
